apps/emqx_retainer/src/emqx_retainer_mnesia.erl

```diff
@@ -258,4 +386,165 @@ is_table_full() ->
+                          write_indices = WriteIndices,
+                          reindexing = Reindexing}] ->
+                          case {ReadIndices, WriteIndices, Reindexing} of
+                              {_, _, true} ->
+                                  ok;
+                              {ConfigIndices, ConfigIndices, false} ->
+                                  ok;
+                              {DBWriteIndices, DBReadIndices, false} ->
+                                  {error, DBWriteIndices, DBReadIndices}
+                          end;
+                      [] ->
+                          mnesia:write(
+                            ?TAB_INDEX_META,
+                            #retained_index_meta{
+                               key = ?META_KEY,
+                               read_indices = ConfigIndices,
+                               write_indices = ConfigIndices,
+                               reindexing = false},
+                            write)
+                  end
+          end,
+    case mria:transaction(?RETAINER_SHARD, Fun) of
+        {atomic, ok} -> ok;
+        {atomic, {error, DBWriteIndices, DBReadIndices}} ->
+            ?SLOG(warning, #{ msg => "emqx_retainer_outdated_indices"
+                            , config_indices => ConfigIndices
+                            , db_write_indices => DBWriteIndices
+                            , db_read_indices => DBReadIndices
+                          }),
+            ok;
+        {aborted, Reason} ->
+            ?SLOG(error, #{ msg => "failed_to_populate_emqx_retainer_indices"
+                            , reason => Reason
+                          }),
+            {error, Reason}
+    end.
+
+db_indices(Type) ->
+    case mnesia:read(?TAB_INDEX_META, ?META_KEY) of
+        [#retained_index_meta{read_indices = ReadIndices, write_indices = WriteIndices}] ->
+            case Type of
+                read -> ReadIndices;
+                write -> WriteIndices
+            end;
+        [] -> []
+    end.
+
+batch_read_number() ->
+    case emqx:get_config([retainer, flow_control, batch_read_number]) of
+        0 -> all_remaining;
+        BatchNum when is_integer(BatchNum) -> BatchNum
+    end.
+
+reindex(NewIndices, Force, StatusFun)
+  when is_boolean(Force) andalso is_function(StatusFun, 1) ->
+
+    %% Disable read indices and update write indices so that new records are written
+    %% with correct indices. Also block parallel reindexing.
+    case try_start_reindex(NewIndices, Force) of
+        {atomic, ok} ->
+            %% Clear old index records.
+            {atomic, ok} = mria:clear_table(?TAB_INDEX),
+
+            %% Fill index records in batches.
+            QH = qlc:q([Topic || #retained_message{topic = Topic} <- ets:table(?TAB_MESSAGE)]),
+            ok = reindex_batch(qlc:cursor(QH), 0, StatusFun),
+
+            %% Enable read indices and unlock reindexing.
+            finalize_reindex();
```

Comment:
If `reindex_batch` returns `{error, _}`, it'll raise an error and reindexing will be locked.

The intended fix is to then run the CLI with `force = true`?

Reply:
Yes, after failure investigation.